ihmc-java-toolkit/src/main/java/us/ihmc/tools/thread/LoopingThread.java

```diff
@@ -0,0 +1,193 @@
+ * Once started, make sure to close this thread.
+ */
+public class LoopingThread extends Thread
+{
+   private final Runnable target;
+   private final Throttler throttler = new Throttler();
+   private volatile double loopPeriodLimit = -1.0;
+   private final AtomicReference<RunCommand> runCommand = new AtomicReference<>(RunCommand.INITIALIZE);
+
+   public LoopingThread() {
+      this(-1.0);
+   }
+
+   public LoopingThread(double loopFrequencyLimit)
+   {
+      super();
+      target = null;
+      limitLoopFrequency(loopFrequencyLimit);
+   }
+
+   public LoopingThread(double loopFrequencyLimit, String name) {
+      this(null, loopFrequencyLimit, name);
+   }
+
+   public LoopingThread(Runnable target, String name)
+   {
+      this(target, -1.0, name);
+   }
+
+   public LoopingThread(Runnable target, double loopFrequencyLimit, String name)
+   {
+      super(target, name);
+      this.target = target;
+      limitLoopFrequency(loopFrequencyLimit);
+   }
+
+   public LoopingThread(RunnableThatThrows targetThatThrows, ExceptionHandler exceptionHandler, String name)
+   {
+      this(targetThatThrows, exceptionHandler, -1.0, name);
+   }
+
+   public LoopingThread(RunnableThatThrows targetThatThrows, ExceptionHandler exceptionHandler, double loopFrequencyLimit, String name)
+   {
+      super(() -> ExceptionTools.handle(targetThatThrows, exceptionHandler), name);
+      target = () -> ExceptionTools.handle(targetThatThrows, exceptionHandler);
+      limitLoopFrequency(loopFrequencyLimit);
+   }
+
+   /**
+    * Limit the frequency of the execution loop.
+    * To un-limit the loop frequency, pass in a number less than or equal to 0.0.
+    * @param frequencyLimit The limit for the loop frequency.
+    *                       If zero or negative, the loop's frequency is not limited.
+    */
+   public void limitLoopFrequency(double frequencyLimit)
+   {
+      loopPeriodLimit = Conversions.hertzToSeconds(frequencyLimit);
+   }
+
+   /**
+    * Starts executing the {@link #runInLoop()} method in a loop.
+    */
+   @Override
+   public void start()
+   {
+      synchronized (runCommand)
+      {
+         initializeIfNeeded();
+
+         runCommand.set(RunCommand.LOOP);
+         runCommand.notify();
+      }
+   }
+
+   /**
+    * Signal the thread to execute the {@link #runInLoop()} method once, then pause.
+    */
+   public void runOnce()
+   {
+      synchronized (runCommand)
+      {
+         initializeIfNeeded();
+
+         runCommand.set(RunCommand.RUN_ONCE);
+         runCommand.notify();
+      }
+   }
+
+   private void initializeIfNeeded()
+   {
+      if (runCommand.compareAndSet(RunCommand.INITIALIZE, RunCommand.PAUSE))
+         super.start();
+   }
+
+   /**
+    * Pauses re-executing the {@link #runInLoop()} method upon its completion.
+    */
+   public void pause()
+   {
+      runCommand.set(RunCommand.PAUSE);
+   }
+
+   /**
+    * Finalizes the execution of {@link #runInLoop()} and stops the thread.
```

Comment:
What does finalize mean here?

Reply:
I was trying to convey that he `runInLoop()` method will finish executing for the last time

Reply:
I think what you just said would be way more clear